File: pcb-maker/inference_pcb.py
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import PeftModel

def load_pcb_model(base_model_path, adapter_path):
    """
    Loads the Allegro X Neural Core (Base Model + LoRA Adapters)
    """
    print(f"Allegro X: Loading Base Model ({base_model_path})...")
    tokenizer = AutoTokenizer.from_pretrained(base_model_path)
    
    model = AutoModelForCausalLM.from_pretrained(
        base_model_path,
        torch_dtype=torch.float16,
        device_map="auto"
    )
    
    print(f"Allegro X: Applying Neural Adapters ({adapter_path})...")
    model = PeftModel.from_pretrained(model, adapter_path)
    
    return model, tokenizer

def generate_pcb_layout(prompt, model, tokenizer):
    """
    Generates PCB layout JSON based on semantic intent
    """
    input_text = f"Intent: {prompt}\nJSON Layout Data:"
    inputs = tokenizer(input_text, return_tensors="pt").to("cuda")
    
    print(f"Allegro X: Running Inference for '{prompt}'...")
    with torch.no_grad():
        outputs = model.generate(
            **inputs, 
            max_new_tokens=512, 
            temperature=0.7, 
            do_sample=True
        )
    
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    return response

if __name__ == "__main__":
    # Example paths
    BASE_MODEL = "codellama/CodeLlama-7b-hf"
    ADAPTER_DIR = "./pcb-llm-model"
    
    # This is a simulation/template. In a real environment, 
    # you would have these models downloaded.
    print("--- Allegro X Generative Neural Core v2.0 ---")
    
    # Sample Mock Output
    user_prompt = "Design a 5v power supply with a 7805 regulator and filtering capacitors"
    
    mock_json_output = """
    {
      "components": [
        {"ref": "J1", "type": "connector", "x": 50, "y": 100},
        {"ref": "U1", "type": "ic", "x": 150, "y": 100, "part": "7805"},
        {"ref": "C1", "type": "capacitor", "x": 100, "y": 80, "val": "10uF"},
        {"ref": "C2", "type": "capacitor", "x": 200, "y": 80, "val": "100nF"}
      ],
      "nets": [
        ["J1.1", "C1.1", "U1.1"],
        ["U1.3", "C2.1", "J2.1"],
        ["J1.2", "C1.2", "U1.2", "C2.2", "J2.2"]
      ]
    }
    """
    
    print(f"Input Prompt: {user_prompt}")
    print("\n[Neural Core Output]")
    print(mock_json_output)

File: pcb-maker/inference_pcb.js
// ===================================================================
// ALLEGRO X -- PCB Inference Script (Node.js Conceptual Port)
// Original: inference_pcb.py (Python + HuggingFace + PEFT)
//
// NOTE: Model loading & inference require Python ML libraries.
// This JS version demonstrates the inference flow and mock output.
// ===================================================================

const BASE_MODEL = "codellama/CodeLlama-7b-hf";
const ADAPTER_DIR = "./pcb-llm-model";

function loadPcbModel(baseModelPath, adapterPath) {
    console.log(`Allegro X: Loading Base Model (${baseModelPath})...`);
    console.log(`Allegro X: Applying Neural Adapters (${adapterPath})...`);
    // In Python: AutoModelForCausalLM + PeftModel
    return { model: "mock", tokenizer: "mock" };
}

function generatePcbLayout(prompt) {
    console.log(`Allegro X: Running Inference for '${prompt}'...`);

    // Mock output (same as Python version)
    const mockJsonOutput = {
        components: [
            { ref: "J1", type: "connector", x: 50, y: 100 },
            { ref: "U1", type: "ic", x: 150, y: 100, part: "7805" },
            { ref: "C1", type: "capacitor", x: 100, y: 80, val: "10uF" },
            { ref: "C2", type: "capacitor", x: 200, y: 80, val: "100nF" }
        ],
        nets: [
            ["J1.1", "C1.1", "U1.1"],
            ["U1.3", "C2.1", "J2.1"],
            ["J1.2", "C1.2", "U1.2", "C2.2", "J2.2"]
        ]
    };

    return mockJsonOutput;
}

// Main execution
console.log("--- Allegro X Generative Neural Core v2.0 ---");
const userPrompt = "Design a 5v power supply with a 7805 regulator and filtering capacitors";

loadPcbModel(BASE_MODEL, ADAPTER_DIR);
const result = generatePcbLayout(userPrompt);

console.log(`\nInput Prompt: ${userPrompt}`);
console.log("\n[Neural Core Output]");
console.log(JSON.stringify(result, null, 2));
